tasks/service/async/src/main/java/org/projectnessie/nessie/tasks/async/vertx/VertxTasksAsync.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2024 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.nessie.tasks.async.vertx;
+
+import io.vertx.core.Vertx;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+import org.projectnessie.nessie.tasks.async.TasksAsync;
+
+public class VertxTasksAsync implements TasksAsync {
+  private final Vertx vertx;
+  private final Clock clock;
+  private final long minimumDelayMillis;
+
+  public VertxTasksAsync(Vertx vertx, Clock clock, long minimumDelayMillis) {
+    this.vertx = vertx;
+    this.clock = clock;
+    this.minimumDelayMillis = minimumDelayMillis;
+  }
+
+  @Override
+  public CompletionStage<Void> call(Runnable runnable) {
+    return supply(
+        () -> {
+          runnable.run();
+          return null;
+        });
+  }
+
+  @Override
+  public <R> CompletionStage<R> supply(Supplier<R> supplier) {
```

Comment:
As pointed offline, the current impl will not interrupt running tasks on cancellation. It's quite tricky to do that.

Taking inspiration from [this conversation](https://groups.google.com/g/vertx/c/byh_qr6a6_s/m/18sYs3E7BQAJ), I came up with the (rather convoluted) solution below:

```java
  @Override
  public <R> CompletionStage<R> supply(Supplier<R> supplier) {
    AtomicBoolean completed = new AtomicBoolean(false);
    AtomicReference<Thread> thread = new AtomicReference<>(null);
    CountDownLatch interrupted = new CountDownLatch(1);
    CompletionStage<R> stage =
        vertx
            .executeBlocking(
                () -> {
                  if (!thread.compareAndSet(null, Thread.currentThread())) {
                    return null;
                  }
                  try {
                    return supplier.get();
                  } catch (Throwable t) {
                    throw new CompletionException(t);
                  } finally {
                    try {
                      if (!completed.compareAndSet(false, true)) {
                        // this task has been cancelled, wait until the interruption attempt is done
                        try {
                          interrupted.await();
                        } catch (InterruptedException ignored) {
                        }
                      }
                    } finally {
                      thread.set(null);
                    }
                  }
                },
                false)
            .toCompletionStage();
    stage.whenComplete(
        (r, t) -> {
          if (t instanceof CancellationException) {
            if (completed.compareAndSet(false, true)) {
              Thread worker = thread.getAndSet(Thread.currentThread());
              if (worker != null) {
                try {
                  worker.interrupt();
                } catch (SecurityException ignored) {
                } finally {
                  interrupted.countDown();
                }
              }
            }
          }
        });
    return stage;
  }

  @Override
  public CompletionStage<Void> schedule(Runnable runnable, Instant scheduleNotBefore) {
    long realDelay = calculateDelay(clock, minimumDelayMillis, scheduleNotBefore);

    // Cannot use Vertx.timer(), because current Quarkus 3.6.5 has a Vertx version that does not
    // have Vertx.timer(). We can use this once Quarkus uses a Vertx version >= 4.5.1:
    //    Timer timer = vertx.timer(realDelay, TimeUnit.MILLISECONDS);
    //    CompletionStage<Void> stage = timer.toCompletionStage().thenRun(runnable);
    //    return new VertxScheduledHandle(timer, stage);

    CompletableFuture<Void> completableFuture = new CompletableFuture<>();
    long timerId =
        vertx.setTimer(
            realDelay,
            id ->
            // scheduling the runnable as a blocking task, a timer handler must not block
            {
              CompletionStage<Void> stage = call(runnable);
              completableFuture.whenComplete(
                  (v, t) -> {
                    if (t instanceof CancellationException) {
                      stage.toCompletableFuture().cancel(true);
                    }
                  });
            });

    completableFuture.whenComplete(
        (v, t) -> {
          if (t instanceof CancellationException) {
            vertx.cancelTimer(timerId);
          }
        });

    return completableFuture;
  }
```

With the above change, the `cancelRunningWithInterrupt` test also works for Vertx. 

But at this point, I'm wondering if we need the Vertx impl at all? Id doesn't seem suitable for long, blocking tasks.

Reply:
Vert.X is still needed for the `TaskServiceImpl` when running in Quarkus.

Reply:
How actual implementations run tasks via `TaskRequest.submitExecution()` is up to them - added notes to the javadoc.

Reply:
I think that it's okay that tasks cannot be interrupted. `TasksServiceImpl` is fine with not interrupting tasks. Other implementations would need to choose the right implementation.